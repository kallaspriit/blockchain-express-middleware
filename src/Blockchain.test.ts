import Axios from "axios";
import MockServer from "axios-mock-adapter";
import * as HttpStatus from "http-status-codes";
import { Blockchain, ILogger, Invoice } from "./";

const CALLBACK_URL = "https://example.com";
const RECEIVING_ADDRESS = "1FupTEd3PDF7HVxNrzNqQGGoWZA4rwiphq";
const API_KEY = "xxx";
const XPUB = "yyy";
const SECRET = "zzz";

let mockServer: MockServer;

describe("Blockchain", () => {
  beforeEach(() => {
    mockServer = new MockServer(Axios);
  });

  afterEach(() => {
    mockServer.restore();
  });

  it("should generate a new receiving address", async () => {
    mockServer.onGet(/receive/).reply(HttpStatus.OK, {
      address: RECEIVING_ADDRESS,
      index: 0,
      callback: CALLBACK_URL,
    });

    const blockchain = new Blockchain({
      apiKey: API_KEY,
      xPub: XPUB,
    });

    const receivingAddress = await blockchain.generateReceivingAddress(CALLBACK_URL);

    expect(receivingAddress).toMatchSnapshot();
  });

  it("should throw error when generating receving address and getting a non 2xx response", async () => {
    mockServer.onGet(/receive/).reply(HttpStatus.BAD_REQUEST, "Bad request");

    const blockchain = new Blockchain({
      apiKey: API_KEY,
      xPub: XPUB,
      gapLimit: 20, // we can provide custom gap limit
    });

    await expect(blockchain.generateReceivingAddress(CALLBACK_URL)).rejects.toMatchSnapshot();
  });

  it("should create a new invoice with receiving address", async () => {
    mockServer.onGet(/receive/).reply(HttpStatus.OK, {
      address: RECEIVING_ADDRESS,
      index: 0,
      callback: CALLBACK_URL,
    });

    const blockchain = new Blockchain({
      apiKey: API_KEY,
      xPub: XPUB,
    });

    const invoice = await blockchain.createInvoice({
      dueAmount: 1,
      message: "Test invoice",
      secret: SECRET,
      callbackUrl: CALLBACK_URL,
    });

    expect(processInvoiceForSnapshot(invoice)).toMatchSnapshot();
  });

  it("should accept custom logger", async () => {
    mockServer.onGet(/receive/).reply(HttpStatus.OK, {
      address: RECEIVING_ADDRESS,
      index: 0,
      callback: CALLBACK_URL,
    });

    const mockLogger: ILogger = {
      trace: jest.fn(),
      debug: jest.fn(),
      info: jest.fn(),
      warn: jest.fn(),
      error: jest.fn(),
    };

    const blockchain = new Blockchain(
      {
        apiKey: API_KEY,
        xPub: XPUB,
      },
      mockLogger,
    );

    await blockchain.createInvoice({
      dueAmount: 2.5,
      message: "Another invoice",
      secret: SECRET,
      callbackUrl: CALLBACK_URL,
    });

    // tslint:disable-next-line:no-any
    expect((mockLogger.info as any).mock.calls).toMatchSnapshot();
  });
});

function processInvoiceForSnapshot(invoice: Invoice): Invoice {
  invoice.createdDate = new Date("2018-04-19T13:48:05.316Z");
  invoice.updatedDate = new Date("2018-04-19T13:48:05.316Z");

  return invoice;
}
