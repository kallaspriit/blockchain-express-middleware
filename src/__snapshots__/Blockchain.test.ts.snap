// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blockchain should accept custom logger 1`] = `
Array [
  Array [
    Object {
      "parameters": Object {
        "callback": "https://example.com?signature=4b506b7c3a37df75af9ac0a54e810ce3397fbe2255d9d820c041f9bec3a3516cfc39bce02570dfed230d81f5f37f0ba5bf886275c69d1e8b31399df45a250dd2",
        "key": "xxx",
        "xpub": "yyy",
      },
      "url": "https://api.blockchain.info/v2/receive?xpub=yyy&callback=https%3A%2F%2Fexample.com%3Fsignature%3D4b506b7c3a37df75af9ac0a54e810ce3397fbe2255d9d820c041f9bec3a3516cfc39bce02570dfed230d81f5f37f0ba5bf886275c69d1e8b31399df45a250dd2&key=xxx",
    },
    "generating receiving address",
  ],
]
`;

exports[`Blockchain should create a new invoice with receiving address 1`] = `
Object {
  "address": "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq",
  "createdDate": 2018-04-19T13:48:05.316Z,
  "dueAmount": 1,
  "message": "Test invoice",
  "paymentState": "PENDING",
  "stateTransitions": Array [],
  "transactions": Array [],
  "updatedDate": 2018-04-20T13:48:05.316Z,
}
`;

exports[`Blockchain should generate a new receiving address 1`] = `
Object {
  "address": "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq",
  "callback": "https://example.com",
  "index": 0,
}
`;

exports[`Blockchain should throw error when generating receving address and getting a non 2xx response 1`] = `[Error: Request failed with status code 400]`;
